core/iwasm/libraries/lib-wasi-threads/lib_wasi_threads_wrapper.c

```diff
@@ -31,14 +40,57 @@ typedef struct {
 static int32
 allocate_thread_id()
 {
-    static int32 thread_id = 0;
-
     int32 id;
 
     os_mutex_lock(&thread_id_lock);
-    id = thread_id++;
+    if (avail_tids.pos < 0) { // Resize stack and push new thread ids
+        uint32 old_size = avail_tids.size;
+        uint32 new_size = avail_tids.size * 2;
+        if (new_size / 2 != avail_tids.size) {
+            LOG_ERROR("Overflow detected during new size calculation");
+            goto resize_fail;
+        }
+        avail_tids.size *= 2;
+
+        size_t realloc_size = avail_tids.size * sizeof(int32);
+        if (realloc_size / sizeof(int32) != avail_tids.size) {
+            LOG_ERROR("Overflow detected during realloc");
+            goto resize_fail;
+        }
+        int32 *tmp =
+            (int32 *)wasm_runtime_realloc(avail_tids.ids, realloc_size);
+        if (avail_tids.ids == NULL) {
+            LOG_ERROR("Thread ID allocator realloc failed");
+            goto resize_fail;
+        }
+        avail_tids.ids = tmp;
+
+        for (uint32 i = 0; i < old_size; i++)
+            avail_tids.ids[i] = avail_tids.size - i;
+
+        avail_tids.pos = old_size - 1;
+    }
+
+    // Pop available thread identifier from `avail_tids` stack
+    id = avail_tids.ids[avail_tids.pos--];
```

Comment:
[minor] if you initialize I'd with -1 you can skip duplication, just rename the resize_fail label to e.g. final.